Add periodic controller to update drop/forward count to prometheus

manalibhutiyani opened the pull request:
Prometheus controller changes for drops/forwards

```release-note
Add periodic controller to update drop/forward count to prometheus
```
 Fixes: #4171
 Signed-off by: Manali Bhutiyani <manali@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4225)
<!-- Reviewable:end -->

Commit:
bpf: Add metricsmap helper funcs and controller metricsmap-bpf-prom-sync

This change adds various helper function to read the metricsmap.
It also adds the functions to iterate through the map and
aggregate the data to prometheus specific counters.

In addition we support the daemon to run a new periodic controller `metricsmap-bpf-prom-sync`
which syncs and aggregates the metrics map with the prometheus server.
The current time frequency of this controller is 5 sec.

Fixes: #4171
Signed-off by: Manali Bhutiyani <manali@covalent.io>

Commit:
pkg/metrics: Add prometheus counterVectors for Drops and Forwards

This change adds new prometheus counter vectors for DropCounts and ForwardCounts
tagged with labels (Drop Reason for drop counts) and direction (ingress/egress)

Fixes: #4171
Signed-off by: Manali Bhutiyani <manali@covalent.io>

rlenglet commented on daemon/daemon.go:
```diff
@@ -871,6 +871,14 @@ func (d *Daemon) init() error {
+		controller.NewManager().UpdateController("metricsmap-bpf-prom-sync",
+			controller.ControllerParams{
+				DoFunc:      func() error { return metricsmap.SyncMetricsMap() },
+				RunInterval: time.Duration(5) * time.Second,
```
The explicit conversion is not necessary:
```
RunInterval: 5 * time.Second,
```

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -51,6 +99,16 @@ func (k *Key) String() string {
 }
 
+// GetDirection gets the direction in human readable string format
+func (k *Key) GetDirection() string {
```
We usually drop the `Get` from getter function names:
GetDirection -> Direction
GetReason -> Reason
GetCount -> Count
GetBytes -> Bytes
GetCountFloat -> CountFloat
GetBytesFloat -> BytesFloat

manalibhutiyani replied:
`Reason` is also the field in `Key` so it does not let me declare a `Reason` function.
I can change this to `DropForwardReason` instead
Same with ` Count` and ` Bytes`

manalibhutiyani replied:
Changing
`Reason` --> `DropForwardReason`
`GetDirection` --> `Direction`
`GetCount`      --> `RequestCount`
`GetBytes`      -->   `RequestBytes`
`GetCountFloat` -> `CountFloat`
`GetBytesFloat` -> `BytesFloat`

Also I noticed other *map.go like `policymap` have actually used `Get*` functions everywhere.

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -32,6 +34,52 @@ const (
+
+// dropForwardReason can be 0 for forwarded or non-zero which specifies
+// the BPF drop reason
+var dropForwardReason = map[uint8]string{
```
There's another map (`errors`) of drop reasons, in `pkg/monitor/datapath_drop.go`.
What prevented you from exporting that map and using it here?

manalibhutiyani replied:
I am not aware of this file, thats all. Let me take a look

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -51,6 +99,16 @@ func (k *Key) String() string {
+	return direction[k.Dir]
+}
+
+// GetReason gets the forwarded/dropped reason in human readable string format
```
Strings like "DROP_UNKNOWN_ICMP_TYPE" are not particularly human readable.
Use the `errors` map defined in `pkg/monitor/datapath_drop.go` instead.

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -59,6 +117,31 @@ func (v *Value) String() string {
 
+// GetCount returns the drop/forward count in a human readable string format
+func (v *Value) GetCount() string {
+	return fmt.Sprintf("%d", v.Count)
```
```
return strconv.FormatUint(v.Count, 10)
```

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -59,6 +117,31 @@ func (v *Value) String() string {
+
+// GetBytes returns drop/forward bytes in a human readable string format
+func (v *Value) GetBytes() string {
+	return fmt.Sprintf("%d", v.Bytes)
```
```
return strconv.FormatUint(v.Bytes, 10)
```

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +169,72 @@ var (
+// and determines which promethues metrics along with respective labels
+// need to be updated.
+func updatePrometheusMetrics(key *Key, val *Value) {
+	if key.IsDrop() {
```
Most of this code can be factorized:
```
var counterVec CounterVec
if key.IsDrop() {
    counterVec = metrics.DropCount
} else {
    counterVec, err = metrics.ForwardCount
}
m, err := counterVec.GetMetricWithLabelValues(key.GetReason(), key.GetDirection())
if err != nil {
    log.Warnf(...)
    return
}
oldValue := metrics.GetCounterValue(m)
newValue := val.GetCountFloat()
// Check if metrics have changed since the last poll.
// If yes, we need to add only the delta.
if newValue > *oldValue {
    metrics.DropCount.WithLabelValues(key.GetReason(), key.GetDirection()).Add((newValue - *oldValue))
}
```

manalibhutiyani replied:
The `DropCount ` and `ForwardCount` take different labels as arguments.

How about this:
```
// UpdatePrometheusMetrics checks the metricsmap key value pair
// and determines which prometheus metrics along with respective labels
// need to be updated.
func UpdatePrometheusMetrics(key *metricsmap.Key, val *metricsmap.Value) {
	var counter prometheus.Counter
	var err error = nil
	if key.IsDrop() {
		counter, err = DropCount.GetMetricWithLabelValues(key.DropForwardReason(), key.Direction())
	} else {
		counter, err = ForwardCount.GetMetricWithLabelValues(key.Direction())
	}
	if err != nil {
		log.Warnf("failed to update prometheus metrics err:",err)
		return
	}
	oldValue := getCounterValue(counter)
	newValue := val.CountFloat()
	// Check if metrics have changed since the last poll.
	// If yes, we need to add only the delta.
	if newValue > oldValue {
		if key.IsDrop() {
			DropCount.WithLabelValues(key.DropForwardReason(), key.Direction()).Add((newValue - oldValue))
		} else {
			ForwardCount.WithLabelValues(key.Direction()).Add((newValue - oldValue))
		}
	}
}

```

rlenglet replied:
Looks good!

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +169,72 @@ var (
+
+	if err != nil {
+		log.WithError(err).WithField(logfields.Path, file).Warn("Unable to open map")
+		return fmt.Errorf("Unable to open metrics map: %s", err)
```
Unable -> unable

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +169,72 @@ var (
+	metricsmap, err := bpf.OpenMap(file)
+
+	if err != nil {
+		log.WithError(err).WithField(logfields.Path, file).Warn("Unable to open map")
```
The log message is too vague:
Unable to open map -> Unable to open metrics map

aanm replied:
Also, @manalibhutiyani check if the logs the error message already.

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +169,72 @@ var (
+
+		entry, err := metricsmap.Lookup(&nextKey)
+		if err != nil {
+			return fmt.Errorf("Unable to lookup metrics map: %s", err)
```
Unable -> unable

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +169,72 @@ var (
+// aggregating it into drops (by drop reason), drops (by direction) and
+// forwards (by direction) with the prometheus server.
+func SyncMetricsMap() error {
+	var file string = bpf.MapPath(MapName)
```
```
file := bpf.MapPath(MapName)
```

manalibhutiyani replied:
done

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -159,6 +160,26 @@ var (
+	DropCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "drop_count_total",
+		Help:      "Total drop requests, tagged by drop reason and direction",
```
To be consistent:
direction -> ingress/egress direction

manalibhutiyani replied:
done

rlenglet commented on pkg/metrics/metrics.go:
```diff
@@ -209,3 +233,15 @@ func SetTSValue(c prometheus.Gauge, ts time.Time) {
+
+// GetCounterValue returns the current value
+// stored for the counter
+func GetCounterValue(m prometheus.Counter) *float64 {
```
Why does this return a pointer?
It never returns a nil pointer, and the pointed value is never modified by the caller, so that's unnecessary.
Return a float64 value instead.

manalibhutiyani replied:
done

rlenglet reviewed (dismissed):
Overall that looks good.

aanm commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -32,6 +34,52 @@ const (
+// direction is the metrics direction i.e ingress (to an endpoint)
+// or egress (from an endpoint).
+var direction = map[uint8]string{
+	1: "INGRESS",
```
correct me if I'm wrong but I thought 0 was INGRESS and 1 was egress?

manalibhutiyani replied:
No, we were having issues with with forward reason which is 0 and in ingress direction . This way the whole map key was 0 and we had trouble looking it up in the bpf map and inserting/updating. Hence this was changed to 1, 2 i.e to make the map key a non zero value.
Thanks

joestringer replied:
FWIW these would be easier to tell if there were constants in this file which determined these directions, with a comment that linked them to the BPF definitions so it's easy to cross-reference their correctness.

aanm commented on pkg/metrics/metrics.go:
```diff
@@ -28,6 +28,7 @@ import (
 
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
```
what is the meaning of `dto`?

rlenglet replied:
DTO is an acronym for Data Transfer Object. That’s from the classic DAO / DTO design pattern. DAO stands for Data Access Object.

Commit:
pkg/monitor: Export DropReason to be consistent with cilium monitor

This commit exports `DropReason` from monitor to the prometheus metrics to
maintain consistency between drop values seen in cilium monitor and the drop values
exported to prometheus.

Fixes: #4171
Signed-off by: Manali Bhutiyani <manali@covalent.io>

Commit:
daemon: Remove unnecessary explicit conversion of time

This commit removes unnecessary explicit conversion of time while starting the
lxcmap-bpf-host-sync controller in the daemon.

Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

rlenglet commented on daemon/daemon.go:
```diff
@@ -868,7 +868,15 @@ func (d *Daemon) init() error {
+		// the prometheus server.
+		controller.NewManager().UpdateController("metricsmap-bpf-prom-sync",
+			controller.ControllerParams{
+				DoFunc:      func() error { return metricsmap.SyncMetricsMap() },
```
This should simply be:
```
DoFunc: metricsmap.SyncMetricsMap,
```

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +133,64 @@ var (
+		counter, err = metrics.ForwardCount.GetMetricWithLabelValues(key.Direction())
+	}
+	if err != nil {
+		log.WithError(err).Warn("failed to update prometheus metrics")
```
failed -> Failed

manalibhutiyani replied:
done

rlenglet commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +133,64 @@ var (
+}
+
+// SyncMetricsMap is called periodically to sync of the metrics map by
+// aggregating it into drops (by drop reason), drops (by direction) and
```
```
... into drops (by drop reason and direction) and
```

manalibhutiyani replied:
done

rlenglet reviewed (approved):
LGTM!
Found 2 more minor nits. Please address them before merging.

tgraf commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -51,6 +63,16 @@ func (k *Key) String() string {
 
+// Direction gets the direction in human readable string format
+func (k *Key) Direction() string {
+	return direction[k.Dir]
```
I would do a switch statement here to avoid eventual panic due to out of bound access on invalid map data.

manalibhutiyani replied:
The data is filled in at BPF so the values are predefined and should be sane, but I can add this check if you like

tgraf replied:
Please do. We may change the format of the BPF map in the future and it can cause values that we don't expect now.

manalibhutiyani replied:
done

tgraf commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -59,6 +81,31 @@ func (v *Value) String() string {
+}
+
+// BytesFloat converts the request bytes to float
+func (v *Value) BytesFloat() float64 {
```
Where is this function used?

manalibhutiyani replied:
We had initially used this to send Bytes metrics too. I can remove this

manalibhutiyani replied:
done

tgraf commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +133,64 @@ var (
+		log.WithError(err).Warn("failed to update prometheus metrics")
+		return
+	}
+	oldValue := metrics.GetCounterValue(counter)
```
How does this behave across agent restarts? What will be the value of the counter after an agent restart?

manalibhutiyani replied:
This will persist across agent restarts (Since it is a counter and Prometheus will never reset it)
The whole point of using a counter was to avoid being reset to zero on agent restart.
@rlenglet

manalibhutiyani replied:
@tgraf : My understanding is that the BPF Maps will persist across restarts. If the in-memory prometheus state is lost due to agent restart, we will simply update it with the value in the BPF maps? Is that the concern?

tgraf replied:
The BPF map is persistent, I assume that the in memory Prometheus counter is not. So we will `Add()` the full delta between `0` and the value in the bpf map to the counter each time the agent restarts.

It would be more accurate to not use a counter for now and just sync the BPF map value to get started.

rlenglet replied:
How is the logic in this PR not accurate?
"Adding a value to the counter" only modifies the counter's in-memory value. And that in-memory value is sampled periodically by Prometheus. "Adding a value to the counter" is not sending anything to Prometheus.

tgraf replied:
The missing piece of information was this: `Any time a counter appears to decrease it’ll be treated as though there was a reset to 0 right after the first data point.` This resolves the mystery for me how this was able to handle both BPF map resets and agent restarts with persistent maps.

tgraf reviewed (changes requested):
Commits must be squashed or order must be fixed as current order breaks git bisect.

joestringer commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -86,6 +133,64 @@ var (
+
+	var key, nextKey Key
+	for {
+		err := metricsmap.GetNextKey(&key, &nextKey)
```
Is there a particular reason not to reuse `Map.Dump()` / `Map.DumpWithCallback()` and the `Metrics` map instance in this package for iterating the map / updating prometheus metrics?

manalibhutiyani replied:
Thanks for the review Joe!. `Map.Dump() / Map.DumpWithCallback()` were mostly used in the `cmd` files. I guess I was seeing the various ways in which code in `pkg/maps/*maps` were iterating bpf maps and did no come across `Map.Dump`. This is definitely an option, but in the interest of time I will refrain from making this change as part of this PR, because this is more of a style change as opposed to correctness of code. I can do it as part of a followup PR. Thanks!

ianvernon replied:
If the map is being dumped without these `Dump` and `DumpWithCallback` functions, you should `RLock` the map which is done in `pkg/bpf/map/DumpWithCallback` to ensure that it is not modified from userspace while accessing it. I don't think this is an issue now because it's only updated in the datapath, but it could be changed in the future.

joestringer reviewed (approved):
I hummed for a bit whether the prometheus dependency inside `pkg/maps/metricsmap/` was logically the right place since I usually think of the `pkg/maps/...` as very specific BPF-bridging code, but in this case the map is explicitly for metrics gathering and I don't see a better place for the prometheus-specific bits now.

Little bit of code duplication that could be avoided, but this could be addressed in subsequent follow-ups. Otherwise BPF bits LGTM

Commit:
docs: update docs with L3/L4 drop/forward metrics

Fixes: #4171
Signed-off by: Manali Bhutiyani <manali@covalent.io>

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -29,6 +29,12 @@ Endpoint
 * ``endpoint_regenerating``: Number of endpoints currently regenerating
 * ``endpoint_regenerations``: Count of all endpoint regenerations that have completed, tagged by outcome
 
+Drops/Forwards (L3/L4)
```
~~Are L7 drops separate?~~
EDIT: Looks like it, based on the descriptions lower in the file.

From out-of-band discussion I was under the impression that the forward counts were either incremented from BPF, or from L7 proxy, depending on what kind of policy the traffic is hitting. This description suggests otherwise. Can you clarify?

manalibhutiyani replied:
Drop /Forward can be independent of policy in L3-L4. Drop due to policy is simply one of the reasons. The L7 proxy on the other hand will *always* have the policy in place and the drops/forwards will be relevant to policy. Does that make it more clear?

joestringer replied:
@manalibhutiyani From drop perspective I think this makes sense. However I'm still confused from the forwarding count. Let me start a fresh comment thread from the relevant line below.

ianvernon reviewed (changes requested):
One change needed w.r.t. locking the map for correctness. Everything else, including docs, LGTM!

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -29,6 +29,12 @@ Endpoint
+Drops/Forwards (L3/L4)
+---------------------
+
+* ``drop_count_total``: Total drop requests, tagged by drop reason and ingress/egress direction
```
Are these "drop requests", or "dropped packets"?

tgraf replied:
Clearly packets

manalibhutiyani replied:
I can change the wording to "packets"

manalibhutiyani replied:
done

joestringer commented on Documentation/configuration/metrics.rst:
```diff
@@ -29,6 +29,12 @@ Endpoint
+---------------------
+
+* ``drop_count_total``: Total drop requests, tagged by drop reason and ingress/egress direction
+* ``forward_count_total``: Total forward requests, tagged by ingress/egress direction
```
I'm still struggling with this direction. Firstly, similar my comments above, "forward requests" means potentially something different from "forwarded requests", and also from a BPF perspective it handles packets, not requests. For argument's sake, let's say this is actually "Total forwarded packets". 

Even looking at "Total forwarded packets" under the "L3/L4" heading, one might draw the conclusion that if this counter doesn't go up, then the packets were never forwarded in the L3/L4 part of processing. However, what's actually happening is that the packets are going to the proxy. Based on the below description, these are explicitly not counted. Furthermore, the counter is also not incremented when the packet comes back from the proxy. In such a case where an L7 policy applies to some set of traffic, the "L3/L4 forwarded" metric is *never incremented*.

https://github.com/cilium/cilium/blob/7281c7e994707b5721a4fffc00b6ff3032ee9daf/bpf/lib/trace.h#L114-L116

I think that this is worth describing somehow here. Perhaps it's a matter of stating that "If L7 policy applies, then the packets will never be counted against ``forwarded_count_total``." ?

There's still an (albeit small) possibility that packets would not hit an L3/L4 drop metric, not hit the L3/L4 forwarded metric (because it goes to the proxy), then proceed to be dropped before it reaches the proxy, perhaps for instance because the packet had invalid checksums or something which we don't detect in L3/L4, but the stack will handle. In that case, no counters would be incremented at all. I wonder whether we should attempt to provide visibility into this kind of thing. AFAICT with the current approach, such packets would be ghosts - never counted in any metric even though they went through Cilium.

tgraf replied:
Even if L7 is applied and the packet is forwarded into an endpoint, the L3 counter should get bumped eventually. Right now, the following scenarios exist:
1. Drop at L3
2. Forward at L3, no L7
3. Drop at L7, no L3 counter incremented
4. Forward at L7, L3 forward counter is incremented

I've considered suggesting to bump the counters on the way to the proxy but it seems weird to bump them twice if the packets eventually make it to the endpoint. What we could consider is to only bump at TO_LXC if it is **not** coming from the proxy.

joestringer replied:
Oh. I was concerned about the case of ingress proxy towards the endpoint, however I now believe that according to your logic above, the packet would hit the ingress `bpf_lxc` prog, see that L7 must occur, send to the proxy, the proxy allows it, then the packet comes back from the proxy into `bpf_netdev` and gets forwarded back into the `bpf_lxc` prog again, this time with the SKIP_PROXY flag set, then eventually will hit `send_trace_notify()` there (same as it would in a pure L3/L4 allow)  and increment the counter. In summary I believe that this case is fine.

However for packets sent through the egress proxy (case 4 above, but for egress part of the datapath), I don't think that the L3 "forwarded" metric is ever incremented:

https://github.com/cilium/cilium/blob/0136423913772516756e46f2469311684002106d/bpf/bpf_netdev.c#L380

https://github.com/cilium/cilium/blob/0136423913772516756e46f2469311684002106d/bpf/lib/l3.h#L122-L129

EDIT: But maybe if the above really is a bug, then we can just split this case out into a separate issue and treat it as a bug to be fixed for 1.1. I don't think this is really relevant to this PR any more.
